File: 0733-flood-fill/0733-flood-fill.cpp
class Solution {
public:
    vector<pair<int, int>> visited;

    bool search(int x, int y)
    {   
        for(auto i: visited)
            if(i.first==x && i.second==y)
            {   
                return false;
            }

        return true;
    }
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        int target=image[sr][sc];
        if(target==color)
            sameFill(image, sr, sc, color, target);
        else
            diffFill(image, sr, sc, color, target);
        return image;
    }

    vector<vector<int>> diffFill(vector<vector<int>>& image, int sr, int sc, int color, int target)
    {
        if(image[sr][sc]==target)
        {   
            visited.push_back(make_pair(sr,sc));
            image[sr][sc]=color;
            if(image.size()>sr-1 && image[0].size()>sc)
                diffFill(image, sr-1, sc, color, target);
            if(image.size()>sr && image[0].size()>sc+1)
                diffFill(image, sr, sc+1, color, target);
            if(image.size()>sr+1 && image[0].size()>sc)
                diffFill(image, sr+1, sc, color, target);
            if(image.size()>sr && image[0].size()>sc-1)
                diffFill(image, sr, sc-1, color, target);
        }

        return image;
    }

    vector<vector<int>> sameFill(vector<vector<int>>& image, int sr, int sc, int color, int target)
    {
        if(image[sr][sc]==target && search(sr,sc))
        {   
            visited.push_back(make_pair(sr,sc));
            image[sr][sc]=color;
            if(image.size()>sr-1 && image[0].size()>sc)
                sameFill(image, sr-1, sc, color, target);
            if(image.size()>sr && image[0].size()>sc+1)
                sameFill(image, sr, sc+1, color, target);
            if(image.size()>sr+1 && image[0].size()>sc)
                sameFill(image, sr+1, sc, color, target);
            if(image.size()>sr && image[0].size()>sc-1)
                sameFill(image, sr, sc-1, color, target);
        }
        
        return image;
    }
};

File: 0733-flood-fill/0733-flood-fill.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        if(image[sr][sc] == color)
            return image;

        int m = image.length;
        int n = image[0].length;
        int[][] result = new int[m][n];
        boolean[][] visited = new boolean[m][n];
        int oldColour = image[sr][sc];

        fill(image, result, visited, sr, sc, color, oldColour, m, n);
          
        return image;
    }

    public static void fill(int[][] image, int[][] result, boolean[][] visited, int sr, int sc, int newColour, int oldColour, int m, int n)
    {   
        image[sr][sc] = newColour;

        if(sr>0 && !visited[sr-1][sc] && image[sr-1][sc]==oldColour)
        {
            visited[sr-1][sc] = true;
            fill(image, result, visited, sr-1, sc, newColour, oldColour, m, n);
        }

        if(sc+1<n && !visited[sr][sc+1] && image[sr][sc+1]==oldColour)
        {
            visited[sr][sc+1] = true;
            fill(image, result, visited, sr, sc+1, newColour, oldColour, m, n);
        }

        if(sr+1<m && !visited[sr+1][sc] && image[sr+1][sc]==oldColour)
        {
            visited[sr+1][sc] = true;
            fill(image, result, visited, sr+1, sc, newColour, oldColour, m, n);
        }

        if(sc>0 && !visited[sr][sc-1] && image[sr][sc-1]==oldColour)
        {
            visited[sr][sc-1] = true;
            fill(image, result, visited, sr, sc-1, newColour, oldColour, m, n);
        }
    }
}